_posts/2021-01-11-pisanie-pluginow-do-intellij.md

```diff
@@ -0,0 +1,324 @@
+---
+layout:    post
+title:     "Wprowadzenie do pisania pluginów w intellij"
+date:      2021-01-11 08:00:00 +0100
+published: true
+author:    bradlinski
+image:     
+tags:
+    - intellij
+    - ide
+    - plugin
+---
+
+Intellij IDEA to obecnie jedno z popularniejszych jeśli nie najpopularniejsze IDE dla Javy. Jedną z jego zalet jest duża baza pluginów, dostarczana przez samo JetBrains jak i społeczność. W tym artykule przedstawię podstawowy tworzenia pluginów dla tej platformy. Większoś z zebranych to informacji(poza elementami związanymi z Javą) ma zastosowanie do dowolnego IDE ze stajni JestBrains.
+
+## Sposoby rozszerzania IDE
+W platformie mamy do dyspozycji trzy podstawowe metody rozszerzenia IDE:
+- akcje - kod uruchamiany w momencie gdy użytkownik wprost kliknie np. w element menu kontekstowego lub paska narzędziowego
+- listenery - pozwalają nasłuchiwać na eventy generowane przez platforme lub inne pluginy
+- extension pointy - interfejsy pozwalające na rozszerzenie konkretnych elementów platformy lub innych pluginów
+
+Dokładnie co jak działa możecie przeczytać w dokumentacji IDE. Ja przedstawię je w praktyce. Stworzymy prosty plugin, który wykorzysta wszystkie opisane wyżej metody. Przykładowy plugin, na podstawie adnotacji, wygeneruje plik tekstowy z prostą dokumentacją dla klas Javy(do tego wykorzystane zostaną akcjie i listenery) oraz pozowli na wyświetlenie tej dokumentacji w ramach `Quick Documentation`(Extension point). Mając klasę:
+
+```java
+
+@Doc("To jest testowa klasa")
+public class TestClass {
+
+    @Doc("To jest testowe pole")
+    private String field;
+
+    @Doc("To jest testowe pole " + Constants.ADDITIONAL_TEXT)
+    private String fieldWithExtraInfo;
+
+}
+```
+
+plugin wygeneruje dokumentację w postaci:
+
+```text
+
+TestClass: To jest testowa klasa
+field: To jest testowe pole
+fieldWithExtraInfo: To jest testowe pole z dodatkowym opisem ze stałej
+
+```
+
+Na githubie znajduje się [repo z pluginem](https://github.com/m87/article-app) oraz [repo z testową aplikacją](https://github.com/m87/article-test-app)
+
+## Baza pluginu
+Przygodę z własnym pluginem rozpoczynamy od stworzenia nowego projektu typu `Intellij Platform Plugin`. Wygenerowana zostanie domyślna struktura katalogów oraz plik XML z manifestem(META-INF/plugin.xml). Zawiera on podstawowe informacje o pluginie. Jego uniklany identyfikator, opis, dane autora itd. W tym pliku definujemy również zależność plguinu. W naszym przypadku musimy zdefiniować trzy:
+
+```xml
+<depends>com.intellij.modules.platform</depends>
+<depends>com.intellij.modules.lang</depends>
+<depends>com.intellij.modules.java</depends>
+```
+- `*.platform` - zawiera interfejsy i klasy potrzebne do implementacji serwisów, akcji, ui, rozszerzeń itd.
```

Comment:
Może troszkę się czepiam, ale UI to jednak skrót, więc napiszmy go wielkimi literami :)